### PR TITLE
Default Projection as Files

### LINKED ISSUE #893: Default Projection as File instead of Env Variables
References EPIC #892

Use case:

Today, our implementation is defaulting projection to environment variables, while the spec requires to project as file. 

Requirements:

Check if this has implication on the documentation side. for all samples we are providing

### PR DESCRIPTION
**Pull request by DhritiShikhar**
### Motivation

- Fixes #893 
- By default, set `bindAsFiles` as `true`.

### Changes

- Makes `BindAsFiles` field a boolean pointer.

### Behaviour 

**[DEFAULT BEHAVIOUR]** If `bindAsFiles` is not set by the user, the operator sets `bindAsFiles` as `true` and the service credentials are injected as files.

If `bindAsFiles` is set as `true` by the user, the service credentials are injected as files.

If `bindAsFiles` is set as `false` by the user,the service credentials are injected as environment variables.

### Manual Testing for Default Behaviour

1. Follow the steps given in this [example](https://github.com/redhat-developer/service-binding-operator/tree/master/examples/nodejs_postgresql) to:
- install DB operator, Service Binding Operator
- Import an application
- Create a DB instance 

2. Create the following `ServiceBindingRequest`:

```
      apiVersion: binding.operators.coreos.com/v1alpha1
      kind: ServiceBinding
      metadata:
          name: binding-request
      spec:
          application:
              name: nodejs-app
              group: apps
              version: v1
              resource: deployments
          services:
          -   group: postgresql.baiju.dev
              version: v1alpha1
              kind: Database
              name: db-demo
```

Notice `bindAsFiles` flag is not set.

3. After ServiceBinding's status condition `Ready` is set to `True`, ssh into the application pod by running command:
`k exec -it <pod_name> -- bash`

E.g. `k exec -it nodejs-app-67bdf56459-vfxqt -- bash`

The command prompt will change. 

4. Execute the following command to find the binding files

```
bash-5.0# cd bindings/binding-request
```

5. Ensure service environment variables are not present in the list of environment variables by executing the following command:
```
bash-5.0# printenv
```

### Acceptance Test

The above scenario is tested in acceptance tests. Execute `make test-acceptance` to run the acceptance tests.

The default behaviour i.e. when `bindAsFiles` is not set by the user is tested in this acceptance test -> https://github.com/redhat-developer/service-binding-operator/pull/907/files#diff-87d32cff8bb83e390f5f95e2d617292442f15ace87b2a6c809ad15a18e98dd58R11

### CODE REVIEW

File: api/v1alpha1/servicebinding_types.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ package v1alpha1
 
 import (
 	"errors"
+
```
Comment (DhritiShikhar) [inline api/v1alpha1/servicebinding_types.go:21]: Extra line added by vscode go code formatting.

File: controllers/reconciler.go [hunk 2]
```diff
@@ -3,6 +3,7 @@ package controllers
 import (
 	"context"
 	"errors"
+
```
Comment (DhritiShikhar) [inline controllers/reconciler.go:6]: Extra line added by vscode go code formatting.

File: api/v1alpha1/servicebinding_types.go [hunk 3]
```diff
@@ -230,7 +231,7 @@ func (sbr ServiceBinding) AsOwnerReference() metav1.OwnerReference {
 }
 
 func (spec *ServiceBindingSpec) NamingTemplate() string {
-	if spec.BindAsFiles && spec.NamingStrategy == "" {
+	if *spec.BindAsFiles && spec.NamingStrategy == "" {
```
Comment (pedjak) [inline api/v1alpha1/servicebinding_types.go:234]: what if `spec.BindAsFiles` is `nil`?

File: controllers/binder.go [hunk 4]
```diff
@@ -158,14 +158,26 @@ func (b *binder) updateSpecVolumes(obj *unstructured.Unstructured) error {
...
 		return nil, err
 	}
+
 	volumes = b.removeVolumes(volumes)
-	if err = unstructured.SetNestedField(obj.Object, volumes, b.getVolumesPath()...); err != nil {
+
+	for _, v := range volumes {
+		volume, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&v)
+		if err != nil {
+			return nil, err
```
Comment (pedjak) [inline controllers/binder.go:172]: how is this change related to the PR scope?

File: controllers/binder.go [hunk 5]
```diff
@@ -399,11 +411,12 @@ func (b *binder) updateContainer(container interface{}) (map[string]interface{},
 	if err != nil {
 		return nil, err
 	}
-	if !b.sbr.Spec.BindAsFiles {
+
+	if b.sbr.Spec.BindAsFiles == nil || !*b.sbr.Spec.BindAsFiles {
```
Comment (pedjak) [inline controllers/binder.go:415]: please extract this boolean expression into a dedicated var, for improved readability

File: controllers/binder.go [hunk 6]
```diff
@@ -399,11 +411,12 @@ func (b *binder) updateContainer(container interface{}) (map[string]interface{},
...
 		c.EnvFrom = b.updateEnvFromList(c.EnvFrom, b.sbr.Status.Secret)
 	}
 	// and adding volume mount entries
-	if b.sbr.Spec.BindAsFiles {
+	if b.sbr.Spec.BindAsFiles != nil && *b.sbr.Spec.BindAsFiles {
```
Comment (pedjak) [inline controllers/binder.go:419]: see above

File: controllers/binder.go [hunk 7]
```diff
@@ -426,12 +439,12 @@ func (b *binder) removeContainer(container interface{}) (map[string]interface{},
 		return nil, err
 	}
 
-	if !b.sbr.Spec.BindAsFiles {
+	if b.sbr.Spec.BindAsFiles != nil && !*b.sbr.Spec.BindAsFiles {
```
Comment (pedjak) [inline controllers/binder.go:442]: see above

File: controllers/binder.go [hunk 8]
```diff
@@ -555,10 +568,11 @@ func (b *binder) update(objs *unstructured.UnstructuredList) ([]*unstructured.Un
 				}
 			}
 		}
-		if b.sbr.Spec.BindAsFiles {
+		if b.sbr.Spec.BindAsFiles != nil && *b.sbr.Spec.BindAsFiles {
```
Comment (pedjak) [inline controllers/binder.go:571]: see above

File: controllers/binder.go [hunk 9]
```diff
@@ -601,12 +615,11 @@ func (b *binder) remove(objs *unstructured.UnstructuredList) error {
 			return err
 		}
 
-		if b.sbr.Spec.BindAsFiles {
+		if b.sbr.Spec.BindAsFiles != nil && *b.sbr.Spec.BindAsFiles {
```
Comment (pedjak) [inline controllers/binder.go:618]: see above

File: controllers/binder.go [hunk 10]
```diff
@@ -601,12 +615,11 @@ func (b *binder) remove(objs *unstructured.UnstructuredList) error {
...
 			if updatedObj, err = b.removeSpecVolumes(updatedObj); err != nil {
 				return err
 			}
 		}
-
```
Comment (pedjak) [inline controllers/binder.go:609]: are these whitespace changes needed?

File: controllers/binder_test.go [hunk 11]
```diff
@@ -267,15 +267,21 @@ func TestBinderNew(t *testing.T) {
...
-
 		binder := newBinder(
 			context.TODO(),
 			f.FakeDynClient(),
 			sbr,
 			&ServiceBindingReconciler{restMapper: testutils.BuildTestRESTMapper()},
 		)
-
 		require.NotNil(t, binder)
+
```
Comment (pedjak) [inline controllers/binder_test.go:277]: how are the changes in this test related to the scope of this PR?

File: test/acceptance/features/bindAppToMultipleServices.feature [hunk 12]
```diff
@@ -39,6 +39,7 @@ Feature: Bind a single application to multiple services
             metadata:
                 name: binding1-myapp-2-sbrs
             spec:
+                bindAsFiles: false
```
Comment (pedjak) [inline test/acceptance/features/bindAppToMultipleServices.feature:42]: do we have now an acceptance where `bindAsFiles` is unset? to check the default behavior?

File: api/v1alpha1/servicebinding_types.go [hunk 13]
```diff
@@ -91,7 +92,7 @@ type ServiceBindingSpec struct {
 	// BindAsFiles makes available the binding values as files in the application's container
 	// See MountPath attribute description for more details.
 	// +optional
-	BindAsFiles bool `json:"bindAsFiles,omitempty"`
+	BindAsFiles *bool `json:"bindAsFiles,omitempty"`
```
Comment (pedjak) [inline api/v1alpha1/servicebinding_types.go:95]: CRD needs to be regenerated

**Review by pedjak — CHANGES_REQUESTED** [review]

File: api/v1alpha1/servicebinding_types.go (hunk 13 above)
Comment (DhritiShikhar) [reply api/v1alpha1/servicebinding_types.go:95]: I see this error when I try to regenerate CRDs:
```
➜  service-binding-operator git:(default-projection) ✗ operator-sdk generate crds   
INFO[0000] Running CRD generator.                       
-: pattern ./pkg/apis/...: lstat ./pkg/apis/: no such file or directory
FATA[0000] error generating CRDs from APIs in pkg/apis: error generating CRD manifests: error running CRD generator: not all generators ran successfully
```

File: test/acceptance/features/bindAppToMultipleServices.feature (hunk 12 above)
Comment (DhritiShikhar) [reply test/acceptance/features/bindAppToMultipleServices.feature:42]: I have mentioned the exact test in which `bindAsFiles` is unset on the PR description

File: controllers/binder.go (hunk 5 above)
Comment (DhritiShikhar) [reply controllers/binder.go]: Done here c80db92

File: controllers/binder.go (hunk 7 above)
Comment (DhritiShikhar) [reply controllers/binder.go]: Done here c80db92

File: controllers/binder.go (hunk 9 above)
Comment (DhritiShikhar) [reply controllers/binder.go]: Done here c80db92

File: api/v1alpha1/servicebinding_types.go (hunk 3 above)
Comment (DhritiShikhar) [reply api/v1alpha1/servicebinding_types.go]: Fixed this c80db92

File: controllers/binder.go (hunk 8 above)
Comment (DhritiShikhar) [reply controllers/binder.go]: Done c80db92

File: controllers/binder.go (hunk 4 above)
Comment (DhritiShikhar) [reply controllers/binder.go]: One of the unit test was failing -> https://github.com/redhat-developer/service-binding-operator/pull/907/files#diff-e3755e7d9c38983d49f099aa4702dd1afe67677cd5193ea7cdcf03af46c15570R302

The reason for failure was that `unstructured.SetNestedField(...)` was expecting an `interface{}` value. This piece of code converts `corev1.Volume` to `interface{}`

This piece of code is executed when there are additional volumes present in the deployment other than the `ServiceBinding` ones. I have also added a test for this scenario.

File: controllers/binder.go (hunk 10 above)
Comment (DhritiShikhar) [reply controllers/binder.go]: Fixed this. I do not add any extra lines in the code.. nor do I remove any line from existing code.

File: controllers/binder.go [hunk 14]
```diff
@@ -158,12 +158,20 @@ func (b *binder) updateSpecVolumes(obj *unstructured.Unstructured) error {
...
 	obj *unstructured.Unstructured,
 ) (*unstructured.Unstructured, error) {
+	var volumesUnstructured []interface{}
 	volumes, err := b.extractSpecVolumes(obj)
 	if err != nil {
 		return nil, err
 	}
 	volumes = b.removeVolumes(volumes)
-	if err = unstructured.SetNestedField(obj.Object, volumes, b.getVolumesPath()...); err != nil {
+	for _, v := range volumes {
```
Comment (DhritiShikhar) [inline controllers/binder.go:167]: The reason for adding this piece of code:

One of the unit test was failing -> https://github.com/redhat-developer/service-binding-operator/pull/907/files#diff-e3755e7d9c38983d49f099aa4702dd1afe67677cd5193ea7cdcf03af46c15570R302

The reason for failure was that unstructured.SetNestedField(...) was expecting an interface{} value. This piece of code converts corev1.Volume to interface{}

This piece of code is executed when there are additional volumes present in the deployment other than the ServiceBinding ones. I have also added a test for this scenario.

File: controllers/binder_test.go [hunk 15]
```diff
@@ -267,45 +267,40 @@ func TestBinderNew(t *testing.T) {
 	})
 
 	t.Run("remove", func(t *testing.T) {
+		fakeMock := mocks.NewFake(t, ns)
```
Comment (DhritiShikhar) [inline controllers/binder_test.go:270]: This unit test tests the volume related piece of code added above -> https://github.com/redhat-developer/service-binding-operator/pull/907/files#r599403982

File: api/v1alpha1/servicebinding_types.go (hunk 13 above)
Comment (pedjak) [reply api/v1alpha1/servicebinding_types.go:95]: with migrating to operator-sdk 1.3, we do this by invoking `make generate`

**Comment by Avni-Sharma** [comment]
On looking at the codecov on the codecov website https://app.codecov.io/gh/redhat-developer/service-binding-operator/compare/907/overview I see the codecov as 63.85  %  and + 2.84% . However, I am not able to infer 40.74% of diff hit (target 61.01%)
What should be the actual codecov then 40.74% or 63.85% ?

**Review by pedjak — APPROVED** [review]
